src/bin/pg_autoctl/pgsetup.c

```diff
@@ -499,7 +498,20 @@ read_pg_pidfile(PostgresSetup *pgSetup, bool pg_is_not_running_is_ok)
 
 	if ((fp = fopen_read_only(pidfile)) == NULL)
 	{
-		if (!pg_is_not_running_is_ok)
+		/*
+		 * Maybe we're attempting to read the file during Postgres start-up
+		 * phase and we just got where the file is replaced, when going from
+		 * standalone backend to full service.
+		 */
+		if (maxRetries > 0)
+		{
+			log_trace("read_pg_pidfile: \"%s\" does not exists [%d]",
+					  pidfile, maxRetries);
+			pg_usleep(250 * 1000); /* wait for 250ms and try again */
+			return read_pg_pidfile(pgSetup, pgIsNotRunningIsOk, maxRetries - 1);
+		}
+
+		if (!pgIsNotRunningIsOk)
```

Comment:
This way if pgIsNotRunningIsOk is false we will first wait `maxRetries * 250ms` before continueing. Is that on on purpose?

Reply:
Yeah, we actually retry after 250ms rather than returning an error immediately. We retry up to maxRetries time before telling to the caller that we're out of luck.